src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py

```diff
@@ -0,0 +1,478 @@
+
+        Args:
+            batteries: What batteries should be used for calculation.
+        """
+        super().__init__(batteries)
+
+        self._metrics = [
+            ComponentMetricId.CAPACITY,
+            ComponentMetricId.SOC_LOWER_BOUND,
+            ComponentMetricId.SOC_UPPER_BOUND,
+        ]
+
+    @classmethod
+    def name(cls) -> str:
+        """Return name of the calculator.
+
+        Returns:
+            Name of the calculator
+        """
+        return "Soc_Capacity"
+
+    @property
+    def battery_metrics(self) -> Dict[int, List[ComponentMetricId]]:
+        """Return what metrics are needed for each battery.
+
+        Returns:
+            Map between battery id and set of required metrics id.
+        """
+        return {bid: self._metrics for bid in self._batteries}
+
+    @property
+    def inverter_metrics(self) -> Dict[int, List[ComponentMetricId]]:
+        """Return what metrics are needed for each inverter.
+
+        Returns:
+            Map between inverter id and set of required metrics id.
+        """
+        return {}
+
+    def calculate(
+        self,
+        metrics_data: Dict[int, ComponentMetricsData],
+        working_batteries: Set[int],
+    ) -> Optional[CapacityMetrics]:
+        """Aggregate the metrics_data and calculate high level metric.
+
+        Missing components will be ignored. Formula will be calculated for all
+        working batteries that are in metrics_data.
+
+        Args:
+            metrics_data: Components metrics data, that should be used to calculate the
+                result.
+            working_batteries: working batteries. These batteries will be used
+                to calculate the result. It should be subset of the batteries given in a
+                constructor.
+
+        Returns:
+            High level metric calculated from the given metrics.
+            Return None if there are no component metrics.
+        """
+        result = CapacityMetrics(
+            timestamp=_MIN_TIMESTAMP, total_capacity=0, bound=Bound(lower=0, upper=0)
+        )
+
+        for battery_id in working_batteries:
+            if battery_id not in metrics_data:
+                continue
+
+            metrics = metrics_data[battery_id]
+
+            capacity = metrics.get(ComponentMetricId.CAPACITY)
+            soc_upper_bound = metrics.get(ComponentMetricId.SOC_UPPER_BOUND)
+            soc_lower_bound = metrics.get(ComponentMetricId.SOC_LOWER_BOUND)
+
+            result.timestamp = max(result.timestamp, metrics.timestamp)
+
+            # All metrics are related so if any is missing then we skip the component.
+            if None in {capacity, soc_lower_bound, soc_upper_bound}:
+                continue
+
+            result.total_capacity += capacity
+            result.bound.upper += capacity * soc_upper_bound
+            result.bound.lower += capacity * soc_lower_bound
+
+        return None if result.timestamp == _MIN_TIMESTAMP else result
```

Comment:
I wonder how expensive are the calculations above, and if it wouldn't make sense to first do a pass over the timestamps to see if it makes sense to calculate anything at all.

I guess now you are optimizing for the case where everything is working normally and what I propose will be optimizing for the case where data is missing, so maybe is not worth it, but sharing it just in case. :)

Reply:
hmm good point, however it is extremely rare that all components of certain type are broken.
It would add extra complexity. And I would (most probably) have to check it anyway because mypy /pylint doesn't see that it is not None. :(